app/services/finance/stripe_charge_service.rb

```diff
@@ -0,0 +1,75 @@
+  end
+
+  protected
+
+  delegate :latest_pending_payment_intent, to: :invoice, allow_nil: true
+
+  def create_payment_intent(amount)
+    response = gateway.purchase(amount.cents, payment_method_id, gateway_options)
+
+    with_payment_intent_data_from(response) do |payment_intent_data|
+      next unless invoice
+
+      payment_intent = invoice.payment_intents.create!(payment_intent_id: payment_intent_data['id'], state: payment_intent_data['status'])
+
+      # For PaymentIntent statuses and corresponding recommended actions see https://stripe.com/docs/payments/accept-a-payment-synchronously
+      # - succeeded                => no additional action > the payment has succeeded
+      # - requires_confirmation    => confirm the payment intent
+      # - requires_action          => check `payment_intent_data['next_action']` for instructions
+      # - requires_payment_method  => do not retry > the payment attempt has failed > ask cardholder to replace card data
+
+      case payment_intent.state
+      when 'succeeded'
+        next
+      when 'requires_confirmation'
+        confirm_payment_intent(payment_intent)
+      end
+    end
+  end
+
+  def confirm_payment_intent(payment_intent)
+    # Passing the gateway option `off_session: false` will cause a `requires_action` status on the payment intent in cases where otherwise it would be `requires_payment_method`.
+    # This happens even when the payment intent has been originally created with `off_session: true` – i.e. Stripe allows us to turn an "off_session" payment intent into an "on_session" one.
+    # Along with the `requires_action` status, the param `next_action.use_stripe_sdk.stripe_js` holds the next-step link to get the transaction authenticated
+    response = gateway.confirm_intent(payment_intent.payment_intent_id, payment_method_id, gateway_options)
+
+    with_payment_intent_data_from(response) do |payment_intent_data|
+      payment_intent_status = payment_intent_data['status']
+      payment_intent.update!(state: payment_intent_status)
+
+      next if payment_intent_status == 'succeeded'
+
+      # Because Stripe won't wrap the response of `/payment_intents/:id/confirm` into an 'error' and the resulting status of the payment intent may mislead ActiveMerchant into thinking it was a success.
+      # See https://github.com/activemerchant/active_merchant/blob/b2f5e89eb383429d47e446f248d7bfe4f95ac3d0/lib/active_merchant/billing/gateways/stripe.rb#L690
+      response.instance_variable_set(:@success, false)
+      response.instance_variable_set(:@message, payment_intent_status.humanize)
+    end
+  end
+
+  def with_payment_intent_data_from(response)
+    payment_intent_data = extract_payment_intent_data_from(response)
+    yield(payment_intent_data) if payment_intent_data.present?
+    response
+  end
+
+  def extract_payment_intent_data_from(response)
+    response_params = response.params
+    return response_params if response.success?
```

Comment:
Gotta fix this. It won't work when the gateway is the old `ActiveMerchant::Billing::StripeGateway`, i.e. when there's no payment intents and it still uses the old Charge API.